### PR TITLE
Add support for markdown-style links

### LINKED ISSUE #1279: sequenceDiagram Links
# sequenceDiagram Links

In searching through the docs and trying this out it seems that many of the diagrams support both anchor and JS "links." However, I can't find how to add a link to sequenceDiagram. Basically I want a sequenceDiagram to have one or more clickable links to static pages.

Are links already implemented for sequenceDiagram? If so what are the basics and I'll do a PR  to add this to the docs?

If links are not part of sequenceDiagrams do you expect this to be difficult to add and if I wanted to PR this in which file would I first poke around to make this change?

Thanks.

### PR DESCRIPTION
**Pull request by ryanwjackson**
## :bookmark_tabs: Summary

Add support for markdown-style links on line messages and notes

Resolves #1279 

## :straight_ruler: Design Decisions

This currently only affects line messages and notes on sequence diagrams

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://mermaid.js.org/community/contributing.html)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://mermaid.js.org/community/contributing.html#update-documentation) is used for all new features.
- [x] :butterfly: If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by bernhardkaindl** [comment]
Works great in the [preview](https://deploy-preview-6158--mermaid-js.netlify.app/edit#pako:eNo9jruOwkAMRX_FcQUSkH6KSEhUaKmggqGwEpPJJrHDPAqE-HeGXaDzuQ_bd6y1YTQY-JpYat501HoarayHruZltVUnBn5Ue6AIpx37kbpmuz_PXIxTMGU5_kur37BS35ZzK6_Osqr-Nhg4OJI-wEU9RMcwdNIXRYELfBfz8bsVAIvZHtmiyWNDvrdo5ZFzlKLub1KjiT7xAr2m1qG50BAypamh-Hn7q04kR9UPP57g408l)!!!

@ryanwjackson, the link in
> "Look at [MermaidJS](https://mermaid.js.org/)"

appears to be slightly below the "Look at" at the preview link above for me.
Do you know if this is expected or can be fixed?

**Comment by ryanwjackson** [comment]
@bernhardkaindl do you have an image of it? I see it as vertically aligned. The underline of the linked text is underneath, but the top and bottom of the text lines up for me. Let me know if I'm missing something.

image

**Comment by DanielSmithGK** [comment]
Hi @ryanwjackson thanks a bunch for implementing!

I've tested in the preview (thanks @bernhardkaindl for pointing this out). And links on arrow messages are working, the alignment also seems as expected (using latest Edge).

`M -->> C: [Test URL](https://www.google.com) `

image

The note link also basically works when it's a "simple" URL:

`note over M: [Note link](https://mermaid.js.org/intro/syntax-reference.html)`

image

However, I've noticed two things:
1. As soon as I introduce an anchor inside a URL, it reverts from interpreting the URL as a link back to regular text, e.g.
  `note over M: [Note link](https://mermaid.js.org/intro/syntax-reference.html#configuration)`<br />
  image
 
2. When you look at the simple URL case, the displayed text / link title is very short, however, the note box seems as wide as the URL is long and gets wider the more characters are in the URL (this seemingly doesn't happen with arrow messages).
  Example for: `note over M: [Note link](https://mermaid.js.org/intro/syntax-reference.html)` <br />
  image <br />
  Example for: `note over M: [Note link](https://mermaid.js.org/intro/syntax-reference.htmlAAAAAAAAAAAAAAAAAAAAAAA)` <br />
  image<br />
  It'd be great if only the necessary displayed text / link title is deciding the note width.

Thanks and regards!
Daniel

**Comment by fornax-a** [comment]
> 1. As soon as I introduce an anchor inside a URL, it reverts from interpreting the URL as a link back to regular text, e.g.
>    `note over M: [Note link](https://mermaid.js.org/intro/syntax-reference.html#configuration)`
>    image
>....
> 
> EDIT: The anchor issue with reverting to text is also true for arrow messages, just checked again: `M -->> C: [Test URL](https://mermaid.js.org/intro/syntax-reference.html#configuration)` image

Hi @DanielSmithGK - Encoding the `#` seems to resolve this problem, though not terribly intuitive for users it might be an acceptable workaround.

`M -->> C: [Test URL](https://mermaid.js.org/intro/syntax-reference.html%23configuration)`

**Comment by ryanwjackson** [comment]
I made some updates to address those issues. Let me know if that resolves them.

**Comment by ryanwjackson** [comment]
@DanielSmithGK any clue who can review this or how to get it merged?

**Comment by awangatsanjel** [comment]
@DanielSmithGK Any progress on this?

**Comment by rngcntr** [comment]
@ryanwjackson Cool addition! Do you think it's possible to apply the same to the participant names in the diagram? I'd imagine it to be very helpful to be able to link documentation pages for each participant.

**Comment by awangatsanjel** [comment]
@rngcntr You can adding link to participant in current version. It is under Actor Menus in the sequence diagram page. 

I really don't understand why this feature didn't get into release after it had been done for 10 months. Mermaid diagram is adopted by AI mostly. Its slow progress of evolution makes me disappointed.

**Comment by rngcntr** [comment]
@awangatsanjel I thought more of a direct link without the need of a drop down, ideally with the capability to reference relative document locations instead of full URLs.
But to be fair, I'm new to this project and I don't know if that's even a feature that is in the scope of mermaid itself or rather implementing parties like GitHub or GitLab.

**Comment by awangatsanjel** [comment]
@rngcntr I am with you on the thought of direct link. Mermaid was started as a diagram drawing tool, however it occasionally used markdown as the descripting language. Since AI adopted it markdown because it is pure text can be processed by the LLM, Mermaid became the default diagram language unless you request other specific diagram format. So, Mermaid is turning to a modeling tool. That the way I have used Mermaid for use. GitHub makes it as extension, it is used a lot in the issues and wiki. So it is evolved from offline diagraming tool to online tool. I don't know if the developers had realized this trends. I hope they can move this feature to the production faster.

**Comment by DanielSmithGK** [comment]
Hi @sidharthv96, 

I've no idea if you could help with this or point us in any direction with this issue here (or maybe tag the correct person, if you know them): The [argos check fails](https://app.argos-ci.com/mermaid/mermaid/builds/3725/143733613) for this PR, and it seems like a false-positive. There are no changes noticeable for us, however, the AI thinks there's something off.

Do you perhaps know if this can be approved manually (or skip it) somehow?

Apologies if you're the wrong person, I've just (desperately) checked the last couple repo activities, perhaps you have some insights how we could advance with this.

Thanks and kind regards!
Daniel

File: cypress/integration/rendering/sequencediagram.spec.js [hunk 1]
```diff
@@ -79,6 +79,15 @@ describe('Sequence diagram', () => {
...
+  it('should render links', () => {
+    imgSnapshotTest(
+      `
+      sequenceDiagram
+      Alice->John: Look at [MermaidJS](https://mermaid.js.org/)
+      John->>Alice: Thanks for the link
+      `
+    );
+  });
   it('should handle different line breaks', () => {
```
Comment (sidharthv96) [inline cypress/integration/rendering/sequencediagram.spec.js:91]: We need more test cases, with the edge cases that have been discussed in previous comments.

File: .changeset/witty-beds-turn.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'mermaid': minor
+---
+
+feat:Added support for markdown-style links in line messages and notes
```
Comment (sidharthv96) [inline .changeset/witty-beds-turn.md:5]:
```suggestion
feat(sequence): Add support for markdown-style links in line messages and notes
```

**Review by sidharthv96 — CHANGES_REQUESTED** [review]
The code, implementation, and documentation is great for this particular feature. The text below is not a reflection of the work you've done, but rather due to a tricky situation mermaid is in regarding markdown support. 

We have an ongoing issue with consistent (and universal) markdown support, mainly around syntax for markdown.

The main issue I see is that we have a [proposed syntax](https://mermaid.js.org/syntax/flowchart.html#markdown-strings) for flowcharts, and some diagrams support markdown without that syntax. So it creates confusion. 

Currently sequence diagrams do not support markdown strings. So adding just URL support, without the other formatting options, and without the markdown syntax, creates more edge cases while we try to unify markdown syntax across the diagrams. 


I will bring this PR for discussion in our next team meeting and give you an update. Thanks for your patience.

**Comment by sidharthv96** [comment]
Standardising markdown support was discussed as the exclusive topic in today's weekly call. 
We will be fixing and merging https://github.com/mermaid-js/mermaid/pull/7057 and then continue expanding support to other diagram types. Please hold further work on this PR, as the approach might change on how labels are drawn, and how we support markdown.

**Comment by mclayton7** [comment]
@sidharthv96 now that #7057 and #7275 have been merged, is work on this PR fair game to support links in markdown?
